### DISCUSSION TITLE
Computing correlation matrix and local observables

### DISCUSSION THREAD

**Discussion by Shovan-Physics** [block-hczhai/block2-preview#122] [Q&A] [answered]
Hi, I want to calculate two-site correlations for the Bose-Hubbard model: https://block2.readthedocs.io/en/latest/tutorial/custom-hamiltonians.html#Bose-Hubbard-Model. I find that `get_npdm` does not work. The following code gives an error:
```
driver.get_npdm(mps, pdm_type=1, npdm_expr="CD", mask=[0,1])

---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
Cell In[13], line 1
----> 1 driver.get_npdm(mps, pdm_type=1, npdm_expr="cd", mask=[0,1])

File /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyblock2/driver/core.py:5555, in DMRGDriver.get_npdm(self, ket, pdm_type, bra, soc, site_type, algo_type, npdm_expr, mask, simulated_parallel, fused_contraction_rotation, cutoff, iprint, max_bond_dim)
   5551     mps.info.deallocate_mutable()
   5553 self.align_mps_center(mbra, mket, max_bond_dim=max_bond_dim)
-> 5555 scheme = bw.b.NPDMScheme(perms)
   5556 opdq = (mbra.info.target - mket.info.target)[0]
   5557 if SymmetryTypes.SU2 in bw.symm_type:

UnboundLocalError: cannot access local variable 'perms' where it is not associated with a value
```
Computing the MPO for each term and using `expectation` is inefficient -- For 64 sites and a bond dimension of 200, computing the matrix takes 4 min (which is 6 times longer than the DMRG sweeps):
```
spcorr_mat = []
for i in range(L):
    row = []
    for j in range(L):
        b = driver.expr_builder().add_term("CD", [i, j], 1)
        op = driver.get_mpo(b.finalize(adjust_order=True, fermionic_ops=""))
        opexp = driver.expectation(mps, op, mps)
        row.append(opexp)
    spcorr_mat.append(row)
```
I suppose it will be much faster if we could shift the orthogonality center of the MPS to `i` and contract the bonds between `i` and `j` (for `i < j`), but I not very familiar with the syntax for this. It would be very helpful if you provide an example that also works for other local operators, such as density-density correlations.

Similarly, I would like to calculate expectations of some local operator (say the number fluctuation) at every site without having to form MPOs for all of them.

**Comment by hczhai** [comment, marked as answer]
Thanks for pointing out the issue. If you update ``block2`` to version ``0.5.3rc18``, the ``DMRGDriver.get_npdm`` can work for custom Hamiltonians and custom operators. For Bose-Hubbard model, one can write

```python3
dm = driver.get_npdm(mps, npdm_expr='CD', fermionic_ops='', mask=[0, 1])[0]
dm = driver.get_npdm(mps, npdm_expr='NN', fermionic_ops='', mask=[0, 1])[0]
```

**Reply by Shovan-Physics** [reply]
Cool, thanks!
